pandas/core/frame.py

```diff
@@ -435,6 +437,32 @@ def __init__(
         if isinstance(data, DataFrame):
             data = data._data
 
+        elif hasattr(data, "__dataframe__"):
+            # construct using dict of numpy arrays
+            # TODO(simonjayhawkins) index, columns, dtype and copy arguments
+            obj = cast("dataframe_protocol.DataFrame", data.__dataframe__)
+
+            def _get_column(col):
+                try:
+                    return col.to_numpy()
+                except NotImplementedError:
+                    return col.to_arrow()
+
+            data = {
+                column_name: _get_column(obj[column_name])
+                for column_name in obj.column_names
+            }
+
+            if not index:
+                try:
+                    index = MultiIndex.from_tuples(obj.row_names)
```

Comment:
hmm, I assumed this would fail if the row names were not tuples.

```python
>>> pd.MultiIndex.from_tuples([[0, 0], [0, 1]])
MultiIndex([(0, 0),
            (0, 1)],
           )
```